packages/gatsby/src/utils/webpack/plugins/static-query-mapper.ts

```diff
@@ -76,390 +59,323 @@ function getRealPath(
+            return
+          }
+          visitedModules.add(module)
 
-          staticQueriesByChunkGroup.set(chunkGroup, staticQueryHashes)
-        })
-      }
+          if (module === asyncRequiresModule) {
+            return
+          }
 
-      // group Slice usage by chunkGroup for ease of use
-      for (const {
-        slices,
-        module: webpackModule,
-      } of webpackModulesUsingSlices) {
-        let chunkGroupsDerivedFromEntrypoints: Array<ChunkGroup> = []
-        for (const chunk of compilation.chunkGraph.getModuleChunksIterable(
-          webpackModule
-        )) {
-          for (const chunkGroup of chunk.groupsIterable) {
-            if (chunkGroup === appEntryPoint) {
-              chunkGroupsDerivedFromEntrypoints.push(chunkGroup)
-            } else {
-              chunkGroupsDerivedFromEntrypoints =
-                chunkGroupsDerivedFromEntrypoints.concat(
-                  getChunkGroupsDerivedFromEntrypoint(chunkGroup, appEntryPoint)
-                )
-            }
+          const component = componentModules.get(module)
+          if (component) {
+            config.onComponent(component)
+            // don't return here yet, as component might be imported by another one, and we want to traverse up until we reach async-requires
           }
-        }
 
-        // loop over all component chunkGroups or global ones
-        chunkGroupsDerivedFromEntrypoints.forEach(chunkGroup => {
-          pageSliceUsageByChunkGroup.set(
-            chunkGroup,
-            mergePreviouslyCollectedSlices(
-              slices,
-              pageSliceUsageByChunkGroup.get(chunkGroup)
-            )
-          )
-        })
-      }
+          if (entryModules.has(module)) {
+            config.onRoot()
+            return
+          }
 
-      // group chunkGroups by componentPaths for ease of use
-      for (const [
-        componentPath,
-        webpackModules,
-      ] of webpackModulesByComponentId) {
-        for (const webpackModule of webpackModules) {
-          for (const chunk of compilation.chunkGraph.getModuleChunksIterable(
-            webpackModule
-          )) {
-            for (const chunkGroup of chunk.groupsIterable) {
-              // When it's a direct import from app entrypoint (async-requires) we know we have the correct chunkGroup
-              if (
-                chunkGroup.name === generateComponentChunkName(componentPath)
-              ) {
-                chunkGroupsWithPageComponents.add(chunkGroup)
-                chunkGroupsByComponentPath.set(componentPath, chunkGroup)
+          const incomingConnections =
+            compilation.moduleGraph.getIncomingConnections(module)
+          for (const connection of incomingConnections) {
+            if (connection.originModule instanceof NormalModule) {
+              const shouldTraverse =
+                connection.dependency?.type !==
+                `harmony export imported specifier`
```

Comment:
not sure if this is correct - ignoring the hard coded string comparison that might be flaky, I am not sure if we should ignore this type of connections/dependencies, but I had very weird results otherwise (like all the static queries were marked as globals, but I did add more boundary after I added this, so probably some tests without this check might prove that it's not needed anymore (?) )